[BUG]kb crash when create mongo/ob with hostnetwork

➜  ~ kbcli version
Kubernetes: v1.27.8-gke.1067004
KubeBlocks: 0.9.0-alpha.7
kbcli: 0.9.0-alpha.2

1. Enable useHostNetwork
```
kbcli addon enable mongodb --set useHostNetwork=true

➜  ~ k get cd mongodb -o yaml
apiVersion: apps.kubeblocks.io/v1alpha1
kind: ClusterDefinition
metadata:
  annotations:
    meta.helm.sh/release-name: kb-addon-mongodb
    meta.helm.sh/release-namespace: kb-system
  creationTimestamp: "2024-03-28T09:22:23Z"
  finalizers:
  - clusterdefinition.kubeblocks.io/finalizer
  generation: 2
  labels:
    app.kubernetes.io/instance: kb-addon-mongodb
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: mongodb
    app.kubernetes.io/version: "5.0"
    config.kubeblocks.io/constraints-mongodb-config-constraints: mongodb-config-constraints
    config.kubeblocks.io/constraints-mongodb-env-constraints: mongodb-env-constraints
    config.kubeblocks.io/tpl-mongodb-config: mongodb5.0-config-template
    config.kubeblocks.io/tpl-mongodb-env: mongodb-env-tpl
    config.kubeblocks.io/tpl-mongodb-metrics-config-new: mongodb-metrics-config-new
    config.kubeblocks.io/tpl-mongodb-scripts: mongodb-scripts
    helm.sh/chart: mongodb-0.9.0
  name: mongodb
  resourceVersion: "2655128"
  uid: 69e1dd20-52ed-48fa-be38-c6944f486452
spec:
  componentDefs:
  - characterType: mongodb
    configSpecs:
    - constraintRef: mongodb-config-constraints
      defaultMode: 256
      keys:
      - mongodb.conf
      name: mongodb-config
      namespace: kb-system
      templateRef: mongodb5.0-config-template
      volumeName: mongodb-config
    - defaultMode: 292
      name: mongodb-metrics-config-new
      namespace: kb-system
      templateRef: mongodb-metrics-config-new
      volumeName: mongodb-metrics-config
    - asEnvFrom:
      - mongodb
      constraintRef: mongodb-env-constraints
      keys:
      - env
      name: mongodb-env
      namespace: kb-system
      templateRef: mongodb-env-tpl
      volumeName: mongodb-env
    consensusSpec:
      followers:
      - accessMode: Readonly
        name: secondary
        replicas: 0
      leader:
        accessMode: ReadWrite
        name: primary
        replicas: 0
      updateStrategy: Serial
    logConfigs:
    - filePathPattern: /data/mongodb/logs/mongodb.log*
      name: running
    name: mongodb
    podSpec:
      containers:
      - command:
        - /scripts/replicaset-setup.sh
        env:
        - name: MONGODB_ROOT_USER
          valueFrom:
            secretKeyRef:
              key: username
              name: $(CONN_CREDENTIAL_SECRET_NAME)
              optional: false
        - name: MONGODB_ROOT_PASSWORD
          valueFrom:
            secretKeyRef:
              key: password
              name: $(CONN_CREDENTIAL_SECRET_NAME)
              optional: false
        name: mongodb
        ports:
        - containerPort: 10
          name: mongodb
          protocol: TCP
        resources: {}
        volumeMounts:
        - mountPath: /data/mongodb
          name: data
        - mountPath: /etc/mongodb/mongodb.conf
          name: mongodb-config
          subPath: mongodb.conf
        - mountPath: /etc/mongodb/keyfile
          name: mongodb-config
          subPath: keyfile
        - mountPath: /scripts
          name: scripts
      - command:
        - /bin/agamotto
        - --config=/opt/conf/metrics-config.yaml
        env:
        - name: MONGODB_ROOT_USER
          valueFrom:
            secretKeyRef:
              key: username
              name: $(CONN_CREDENTIAL_SECRET_NAME)
              optional: false
        - name: MONGODB_ROOT_PASSWORD
          valueFrom:
            secretKeyRef:
              key: password
              name: $(CONN_CREDENTIAL_SECRET_NAME)
              optional: false
        image: infracreate-registry.cn-zhangjiakou.cr.aliyuncs.com/apecloud/agamotto:0.1.2-beta.1
        imagePullPolicy: IfNotPresent
        name: metrics
        ports:
        - containerPort: 20
          name: http-metrics
          protocol: TCP
        - containerPort: 21
          name: http-telemetry
          protocol: TCP
        resources: {}
        securityContext:
          runAsNonRoot: true
          runAsUser: 1001
        volumeMounts:
        - mountPath: /opt/conf
          name: mongodb-metrics-config
      dnsPolicy: ClusterFirstWithHostNet
      hostNetwork: true
    probes:
      roleProbe:
        failureThreshold: 3
        periodSeconds: 2
        timeoutSeconds: 2
      roleProbeTimeoutAfterPodsReady: 300
    scriptSpecs:
    - defaultMode: 365
      name: mongodb-scripts
      namespace: kb-system
      templateRef: mongodb-scripts
      volumeName: scripts
    service:
      ports:
      - name: mongodb
        port: 27017
        protocol: TCP
        targetPort: 0
    switchoverSpec:
      withCandidate:
        cmdExecutorConfig:
          command:
          - /scripts/switchover-with-candidate.sh
          image: infracreate-registry.cn-zhangjiakou.cr.aliyuncs.com/apecloud/mongo:5.0.14
        scriptSpecSelectors:
        - name: mongodb-scripts
      withoutCandidate:
        cmdExecutorConfig:
          command:
          - /scripts/switchover-without-candidate.sh
          image: infracreate-registry.cn-zhangjiakou.cr.aliyuncs.com/apecloud/mongo:5.0.14
        scriptSpecSelectors:
        - name: mongodb-scripts
    volumeTypes:
    - name: data
      type: data
    workloadType: Consensus
  connectionCredential:
    endpoint: $(SVC_FQDN):$(SVC_PORT_mongodb)
    headlessEndpoint: $(KB_CLUSTER_COMP_NAME)-0.$(HEADLESS_SVC_FQDN):$(SVC_PORT_mongodb)
    headlessHost: $(KB_CLUSTER_COMP_NAME)-0.$(HEADLESS_SVC_FQDN)
    headlessPort: $(SVC_PORT_mongodb)
    host: $(SVC_FQDN)
    password: $(RANDOM_PASSWD)
    port: $(SVC_PORT_mongodb)
    username: root
  type: mongodb
status:
  observedGeneration: 2
  phase: Available
```
2. Create cluster , kb crash
```
 `kbcli cluster create  mongo-exizum --termination-policy=DoNotTerminate --monitoring-interval=0 --cluster-definition=mongodb --enable-all-logs=false --cluster-version=mongodb-4.4 --set cpu=100m,memory=0.5Gi,replicas=3,storage=3Gi  --namespace default `

Cluster mongo-exizum created

➜  ~ k get pod -n kb-system | grep kubeblocks
kubeblocks-55865d5977-2hqrf                        0/1     CrashLoopBackOff   13 (5m5s ago)    7h30m
kubeblocks-dataprotection-6c99fdcb77-s4mcs         1/1     Running            0                7h30m



2024-03-28T09:35:51.154Z	ERROR	Reconciler error	{"controller": "componentdefinition", "controllerGroup": "apps.kubeblocks.io", "controllerKind": "ComponentDefinition", "ComponentDefinition": {"name":"mongodb"}, "namespace": "", "name": "mongodb", "reconcileID": "516a34ff-b938-43a4-a3d1-3d732f83bc3b", "error": "illegal version string \"5.0\""}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.2/pkg/internal/controller/controller.go:324
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.2/pkg/internal/controller/controller.go:265
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.2/pkg/internal/controller/controller.go:226
2024-03-28T09:35:51.163Z	INFO	KubeAPIWarningLogger	unknown field "spec.template.metadata.creationTimestamp"
2024-03-28T09:35:51.165Z	INFO	**************** commit done	{"controller": "replicatedstatemachine", "controllerGroup": "workloads.kubeblocks.io", "controllerKind": "ReplicatedStateMachine", "ReplicatedStateMachine": {"name":"pulsar-qbyolq-pulsar-broker","namespace":"default"}, "namespace": "default", "name": "pulsar-qbyolq-pulsar-broker", "reconcileID": "17797007-ae93-45d3-8fa5-4e22bc5d4d53", "ReplicatedStateMachine": {"name":"pulsar-qbyolq-pulsar-broker","namespace":"default"}}
2024-03-28T09:35:51.172Z	INFO	**************** prepare	{"controller": "replicatedstatemachine", "controllerGroup": "workloads.kubeblocks.io", "controllerKind": "ReplicatedStateMachine", "ReplicatedStateMachine": {"name":"pulsar-qbyolq-pulsar-broker","namespace":"default"}, "namespace": "default", "name": "pulsar-qbyolq-pulsar-broker", "reconcileID": "20cd065f-8381-42fa-ba95-baca3f8f738e", "ReplicatedStateMachine": {"name":"pulsar-qbyolq-pulsar-broker","namespace":"default"}}
2024-03-28T09:35:51.208Z	INFO	Observed a panic in reconciler: assignment to entry in nil map	{"controller": "component", "controllerGroup": "apps.kubeblocks.io", "controllerKind": "Component", "Component": {"name":"mongo-exizum-mongodb","namespace":"default"}, "namespace": "default", "name": "mongo-exizum-mongodb", "reconcileID": "26eabc3a-f9b3-4b6f-b6bf-f5d4cae39017"}
panic: assignment to entry in nil map [recovered]
	panic: assignment to entry in nil map

goroutine 850 [running]:
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.2/pkg/internal/controller/controller.go:115 +0x1e5
panic({0x2708520?, 0x3133fb0?})
	/usr/local/go/src/runtime/panic.go:914 +0x21f
github.com/apecloud/kubeblocks/controllers/apps.allocateHostPortsWithFunc(0x26d6f80?, 0xc0036be300, 0xc002d69678)
	/src/controllers/apps/transformer_component_hostnetwork.go:94 +0x333
github.com/apecloud/kubeblocks/controllers/apps.allocateHostPorts(0xc0036be300)
	/src/controllers/apps/transformer_component_hostnetwork.go:80 +0x18f
github.com/apecloud/kubeblocks/controllers/apps.(*componentHostNetworkTransformer).Transform(0xc00143f870?, {0x3162f80?, 0xc0019daab0?}, 0xc00143f910?)
	/src/controllers/apps/transformer_component_hostnetwork.go:34 +0x65
github.com/apecloud/kubeblocks/pkg/controller/graph.TransformerChain.ApplyTo({0xc002d4c280?, 0x14, 0x2808ca0?}, {0x3162f80, 0xc0019daab0}, 0x417012?)
	/src/pkg/controller/graph/transformer.go:57 +0x9b
github.com/apecloud/kubeblocks/controllers/apps.(*componentPlanBuilder).Build(0xc0026d8820)
	/src/controllers/apps/component_plan_builder.go:120 +0xb6
github.com/apecloud/kubeblocks/controllers/apps.(*ComponentReconciler).Reconcile(0xc000ca4570, {0x31632d8, 0xc002d2bb60}, {{{0xc0016faf30, 0x7}, {0xc0017474a0, 0x14}}})
	/src/controllers/apps/component_controller.go:167 +0x9f1
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x31632d8?, {0x31632d8?, 0xc002d2bb60?}, {{{0xc0016faf30?, 0x2584980?}, {0xc0017474a0?, 0xc00156b7a0?}}})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.2/pkg/internal/controller/controller.go:118 +0xb7
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc000cba1e0, {0x3163310, 0xc000a7e8c0}, {0x2837e60?, 0xc00156b7a0?})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.2/pkg/internal/controller/controller.go:314 +0x368
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc000cba1e0, {0x3163310, 0xc000a7e8c0})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.2/pkg/internal/controller/controller.go:265 +0x1af
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.2/pkg/internal/controller/controller.go:226 +0x79
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2 in goroutine 119
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.2/pkg/internal/controller/controller.go:222 +0x565
```
➜  ~ kbcli report cluster   --with-logs --all-containers  mongo-exizum
reporting cluster information to report-cluster-mongo-exizum-2024-03-28-17-41-46.zip
processing manifests                               OK
processing events                                  OK
process pod logs                                   OK
➜  ~ kbcli report kubeblocks --with-logs --all-containers --output yaml
reporting KubeBlocks information to report-kubeblocks-2024-03-28-17-41-59.zip
processing manifests                               OK
processing events                                  OK
process pod logs                                   OK
[report-kubeblocks-2024-03-28-17-41-59.zip](https://github.com/apecloud/kubeblocks/files/14787263/report-kubeblocks-2024-03-28-17-41-59.zip)
[report-cluster-mongo-exizum-2024-03-28-17-41-46.zip](https://github.com/apecloud/kubeblocks/files/14787265/report-cluster-mongo-exizum-2024-03-28-17-41-46.zip)

➜  ~ kbcli version
Kubernetes: v1.27.8-gke.1067004
KubeBlocks: 0.9.0-beta.5
kbcli: 0.9.0-beta.1

When create ob with hostnetwork, kb also crashed

 kubectl apply -f cluster.yaml
```
apiVersion: apps.kubeblocks.io/v1alpha1
kind: Cluster
metadata:
  name: oceanbase-cluster
  namespace: default
  annotations:
    "kubeblocks.io/extra-env": "{\"ZONE_COUNT\":\"3\",\"OB_CLUSTERS_COUNT\":\"1\"}"
spec:
  clusterDefinitionRef: ob-ce
  clusterVersionRef: ob-ce-4.2.0.0-100010032023083021
  terminationPolicy: Delete
  componentSpecs:
    - name: ob-ce
      componentDefRef: ob-ce-hostnetwork
      replicas: 3
      resources:
        limits:
          cpu: "4"
          memory: "8Gi"
        requests:
          cpu: "4"
          memory: "8Gi"
      volumeClaimTemplates:
        - name: data-file
          spec:
            accessModes:
              - ReadWriteOnce
            resources:
              requests:
                storage: 50Gi
        - name: data-log
          spec:
            accessModes:
              - ReadWriteOnce
            resources:
              requests:
                storage: 50Gi
        - name: log
          spec:
            accessModes:
              - ReadWriteOnce
            resources:
              requests:
                storage: 10Gi
        - name: workdir
          spec:
            accessModes:
              - ReadWriteOnce
            resources:
              requests:
                storage: 20Gi


0411 03:57:08.891089       1 warnings.go:70] Due to the lack of practical use cases, this API is deprecated from KB 0.9.0.
2024-04-11T03:57:08.921Z	INFO	Observed a panic in reconciler: assignment to entry in nil map	{"controller": "component", "controllerGroup": "apps.kubeblocks.io", "controllerKind": "Component", "Component": {"name":"oceanbase-cluster-ob-ce","namespace":"default"}, "namespace": "default", "name": "oceanbase-cluster-ob-ce", "reconcileID": "3ace1448-b5c1-4899-95f6-05ce38b19852"}
panic: assignment to entry in nil map [recovered]
	panic: assignment to entry in nil map

goroutine 814 [running]:
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.5/pkg/internal/controller/controller.go:116 +0x1e5
panic({0x27b2d20?, 0x320ef80?})
	/usr/local/go/src/runtime/panic.go:914 +0x21f
github.com/apecloud/kubeblocks/controllers/apps.allocateHostPortsWithFunc(0x2780b80?, 0xc00233ec00, 0xc002133658)
	/src/controllers/apps/transformer_component_hostnetwork.go:113 +0x333
github.com/apecloud/kubeblocks/controllers/apps.allocateHostPorts(0xc00233ec00)
	/src/controllers/apps/transformer_component_hostnetwork.go:99 +0x18f
github.com/apecloud/kubeblocks/controllers/apps.(*componentHostNetworkTransformer).Transform(0xc0014f3ae0?, {0x323ec10?, 0xc000f721b0?}, 0xc0014f3b60?)
	/src/controllers/apps/transformer_component_hostnetwork.go:53 +0x65
github.com/apecloud/kubeblocks/pkg/controller/graph.TransformerChain.ApplyTo({0xc001446f00?, 0x14, 0x28b88c0?}, {0x323ec10, 0xc000f721b0}, 0x4178b2?)
	/src/pkg/controller/graph/transformer.go:57 +0x9b
github.com/apecloud/kubeblocks/controllers/apps.(*componentPlanBuilder).Build(0xc0004dd400)
	/src/controllers/apps/component_plan_builder.go:120 +0xb6
github.com/apecloud/kubeblocks/controllers/apps.(*ComponentReconciler).Reconcile(0xc000541740, {0x323efe8, 0xc0020b7e90}, {{{0xc000df7c90, 0x7}, {0xc000559d10, 0x17}}})
	/src/controllers/apps/component_controller.go:167 +0xa11
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x3245180?, {0x323efe8?, 0xc0020b7e90?}, {{{0xc000df7c90?, 0xb?}, {0xc000559d10?, 0x0?}}})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.5/pkg/internal/controller/controller.go:119 +0xb7
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc000650320, {0x323f020, 0xc0004e9e50}, {0x28e8bc0?, 0xc000a87c60?})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.5/pkg/internal/controller/controller.go:316 +0x3cc
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc000650320, {0x323f020, 0xc0004e9e50})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.5/pkg/internal/controller/controller.go:266 +0x1af
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.5/pkg/internal/controller/controller.go:227 +0x79
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2 in goroutine 133
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.5/pkg/internal/controller/controller.go:223 +0x565
➜  ~
```

--- a/controllers/apps/transformer_component_hostnetwork.go
+++ b/controllers/apps/transformer_component_hostnetwork.go
@@ -102,6 +102,12 @@ func allocateHostPorts(synthesizedComp *component.SynthesizedComponent) (map[str
 func allocateHostPortsWithFunc(pm *intctrlutil.PortManager, synthesizedComp *component.SynthesizedComponent,
 	needAllocate func(string, string, int32) bool) (map[string]map[string]int32, error) {
 	ports := map[string]map[string]int32{}
+	insert := func(c, pk string, pv int32) {
+		if _, ok := ports[c]; !ok {
+			ports[c] = map[string]int32{}
+		}
+		ports[c][pk] = pv
+	}
 	for _, c := range synthesizedComp.PodSpec.Containers {
 		for _, p := range c.Ports {
 			portKey := intctrlutil.BuildHostPortName(synthesizedComp.ClusterName, synthesizedComp.Name, c.Name, p.Name)
@@ -110,7 +116,7 @@ func allocateHostPortsWithFunc(pm *intctrlutil.PortManager, synthesizedComp *com
 				if err != nil {
 					return nil, err
 				}
-				ports[c.Name][p.Name] = port
+				insert(c.Name, p.Name, port)
 			} else {
 				if err := pm.UsePort(portKey, p.ContainerPort); err != nil {
 					return nil, err

